src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs

```diff
@@ -660,6 +660,12 @@ private protected sealed override bool IsStringProperty(string memberName)
             return false;
         }
 
+        internal static void DecodeModuleInitializerAttribute(ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
```

Comment:
>DecodeModuleInitializerAttribute [](start = 29, length = 32)

It doesn't look like there is anything to decode. In cases like this we do not add methods like that.

Reply:
This method gets quite long in the branch where I implement diagnostics: https://github.com/dotnet/roslyn/pull/43301/files#diff-b9a132a61804f82152a0f38aab1f9f47R665-R732

Can this be merged as-is with the understanding that `// PROTOTYPE(module-initializers): diagnostics` gets expanded to 68 lines?

Reply:
> This method gets quite long in the branch where I implement diagnostics: [https://github.com/dotnet/roslyn/pull/43301/files#diff-b9a132a61804f82152a0f38aab1f9f47R665-R732](https://github.com/dotnet/roslyn/pull/43301/files#diff-b9a132a61804f82152a0f38aab1f9f47R665-R732)

I see no benefit to have this method in this type, it is better to not mix decoding of what is in attribute with its semantic analysis. For simple cases like this (nothing to decode, no args to decode, etc) we do all the processing inside ```DecodeWellKnownAttributeAppliedToMethod```, including all error reporting. Especially that some errors are context depended, i.e. when the attribute is applied to a non-static or an inaccessible method, etc. It is fine to extract the code into a local function within ```DecodeWellKnownAttributeAppliedToMethod``` or a private helper next to it.

Regarding diagnostics, I believe the rules are rather simple and there aren't that many of them. So, I am surprised to hear that the code is going to be that complex. But I might be wrong.  


---
In reply to: [409486138](https://github.com/dotnet/roslyn/pull/43281#discussion_r409486138) [](ancestors = 409486138)